File: postgresql/golang/utils/postgres_util.go
package utils

import (
	"context"
	"database/sql"
	"log"
	"time"

	"github.com/jmoiron/sqlx"
	_ "github.com/lib/pq"
)

type PostgresUtil interface {
	GetDb() *sqlx.DB
	BeginTxx(ctx context.Context, options *sql.TxOptions) (*sqlx.Tx, error)
	Close(host string, port string)
	CommitOrRollback(tx *sqlx.Tx, err error) error
}

type postgresUtil struct {
	db *sqlx.DB
}

func NewPostgresUtil(host string, username string, password string, database string, port string, applicationName string, maxOpenConnection int, maxIdleConnection int, connectionMaxIdletime int, connectionMaxLifetime int) PostgresUtil {
	println(time.Now().String(), "postgres: connecting to", host, ":", port)
	db, err := sqlx.Connect("postgres", "user="+username+" dbname="+database+" sslmode=disable password="+password+" host="+host+" port="+port+" application_name="+applicationName)
	if err != nil {
		log.Fatalln("postgres: error when connecting to:", err)
	}

	db.SetMaxOpenConns(maxOpenConnection)
	db.SetMaxIdleConns(maxIdleConnection)
	db.SetConnMaxLifetime(time.Duration(connectionMaxIdletime) * time.Minute)
	db.SetConnMaxLifetime(time.Duration(connectionMaxLifetime) * time.Minute)
	println(time.Now().String(), "postgres: connected to", host, ":", port)

	println(time.Now().String(), "postgres: pinging to", host, ":", port)
	err = db.Ping()
	if err != nil {
		log.Fatalln("postgres: error when pinging to:", err)
	}
	println(time.Now().String(), "postgres: pinged to", host, ":", port)

	return &postgresUtil{
		db: db,
	}
}

func (util *postgresUtil) GetDb() *sqlx.DB {
	return util.db
}

func (util *postgresUtil) BeginTxx(ctx context.Context, options *sql.TxOptions) (*sqlx.Tx, error) {
	return util.db.BeginTxx(ctx, options)
}

func (util *postgresUtil) Close(host string, port string) {
	println(time.Now().String(), "postgres: closing to", host, ":", port)
	err := util.db.Close()
	if err != nil {
		log.Fatalln("postgres: error when closing to:", err)
	}
	println(time.Now().String(), "postgres: closed to", host, ":", port)
}

func (util *postgresUtil) CommitOrRollback(tx *sqlx.Tx, err error) error {
	if err == nil {
		err = tx.Commit()
		if err != nil && err != sql.ErrTxDone {
			err = tx.Rollback()
			if err != nil && err != sql.ErrTxDone {
				return err
			}
			return nil
		}
		return nil
	} else {
		err = tx.Rollback()
		if err != nil && err != sql.ErrTxDone {
			return err
		}
		return nil
	}
}

File: postgresql/rust/src/utils/postgres_util.rs
use std::time::Duration;

use chrono::Local;
use sqlx::{postgres::PgPoolOptions, Pool, Postgres, Transaction};

pub trait PostgresUtil {
    async fn get_pool(&self) -> &Pool<Postgres>;
    async fn begin(&self) -> Result<Transaction<'static, Postgres>, sqlx::Error>;
    async fn commit(&self, tx: Transaction<'static, Postgres>) -> Result<(), sqlx::Error>;
    async fn rollback(&self, tx: Transaction<'static, Postgres>) -> Result<(), sqlx::Error>;
    async fn close(&self);
}

#[derive(Debug)]
pub struct PostgresUtilImpl {
    pool: Pool<Postgres>
}

impl PostgresUtilImpl {
    pub async fn new() -> PostgresUtilImpl {
        println!("{} postgres: connecting to localhost:5432", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string());
        let postgres_url = format!("postgres://postgres:12345@localhost:5432/test1?application_name=test1");
        let pool = PgPoolOptions::new()
            .max_connections(10)
            .min_connections(5)
            .acquire_timeout(Duration::from_secs(10))
            .max_lifetime(Duration::from_secs(600))
            .idle_timeout(Duration::from_secs(300))
            .connect(&postgres_url)
            .await.expect("cannot connect to localhost:5432");
        println!("{} postgres: connected to localhost:5432", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string());
        PostgresUtilImpl {
            pool: pool
        }

    }
}

impl PostgresUtil for PostgresUtilImpl {
    async fn get_pool(&self) -> &Pool<Postgres> {
        &self.pool
    }

    async fn begin(&self) -> Result<Transaction<'static, Postgres>, sqlx::Error> {
        self.pool.begin().await
    }

    async fn commit(&self, tx: Transaction<'static, Postgres>) -> Result<(), sqlx::Error> {
        tx.commit().await
    }

    async fn rollback(&self, tx: Transaction<'static, Postgres>) -> Result<(), sqlx::Error> {
        tx.rollback().await
    }

    async fn close(&self) {
        println!("{} postgres: closing to localhost:5432", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string());
        self.pool.close().await;
        println!("{} postgres: closed to localhost:5432", Local::now().format("%Y-%m-%d %H:%M:%S%.3f").to_string());
    }
}